### ISSUE TITLE
[BUG] __delitem__ triggers -Wunused-variable

### ISSUE THREAD

**Issue by wkschwartz** [cython/cython#4032]
# Description
Defining a [`cdef class`](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html) with a [`__delitem__`](https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#sequences-and-mappings) method triggers [`-Wunused-variable`](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-variable).

# To Reproduce

`delitem_test.pyx`:
```cython
cdef class K:
	cdef list a
	def __cinit__(self, list a):                   self.a = list(a)
	def __getitem__(self, Py_ssize_t i):           return self.a[i]
	def __setitem__(self, Py_ssize_t i, object v): self.a[i] = v
	def __delitem__(self, Py_ssize_t i):           self.a[i] = None
```

Compilation:
```console
$ cython -3 delitem_test.pyx     
$ clang -c delitem_test.c -Wunused-variable -I$(python3.7 -c'import sysconfig;print(sysconfig.get_path("include"))')
delitem_test.c:2287:18: warning: unused variable 'o_type_name' [-Wunused-variable]
  __Pyx_TypeName o_type_name;
                 ^
1 warning generated.
$ gcc-10 -c delitem_test.c -Wunused-variable -I$(python3.7 -c'import sysconfig;print(sysconfig.get_path("include"))')
delitem_test.c: In function '__pyx_mp_ass_subscript_12delitem_test_K':
delitem_test.c:2287:18: warning: unused variable 'o_type_name' [-Wunused-variable]
 2287 |   __Pyx_TypeName o_type_name;
      |                  ^~~~~~~~~~~
```

# Expected behavior

The problematic code in the generated `delitem_test.c` is this, starting on line 2286:

```c
static int __pyx_mp_ass_subscript_12delitem_test_K(PyObject *o, PyObject *i, PyObject *v) {
  __Pyx_TypeName o_type_name;
  if (v) {
    return __pyx_pw_12delitem_test_1K_5__setitem__(o, i, v);
  }
  else {
    return __pyx_pw_12delitem_test_1K_7__delitem__(o, i);
  }
}
```

Simply not generating the function's first line declaring `o_type_name` would resolve the warning.

# Environment
- **OS**: macOS 10.15
- **Python**: CPython 3.7.10
- **Cython**: 3.0a6
- **clang**: 12.0.0
- **GCC**: 10.2.0

**Comment by da-woods** [comment]
It looks like this is fixed somewhere between 3.0a6 and the current master. But I'm not sure exactly where.

**Comment by da-woods** [comment]
(For what it's worth it was https://github.com/cython/cython/commit/206adb437d78333f2876da6ffd0dca7258b5ab11#diff-1c3b7395c96b4ecec571dc9f3fb5f1689faa4be6006abe7c3f01145f1010cf9c that fixed it)
